File: frontend/src/utilities/mergeSort.js
const mergeSort = async (
  array,
  updateArray,
  highlightIndices,
  delay,
  updateSkip,
  setSwapCount,
  setSortTime,
  setArrayAccesses,
  setComparisonCount
) => {
  const start = performance.now();
  let swapCount = 0; // No swaps in merge sort, but we can track the movement of elements
  let comparisonCount = 0;
  let arrayAccess = 0;

  let updateCounter = 0; // Counter to control throttling

  const merge = async (left, right, leftStartIndex) => {
    let sorted = [];
    let i = 0; // Pointer for left
    let j = 0; // Pointer for right

    while (i < left.length && j < right.length) {
      highlightIndices([leftStartIndex + i, leftStartIndex + left.length + j]); // Highlight current comparison

      comparisonCount++;
      setComparisonCount(comparisonCount);

      if (left[i] <= right[j]) {
        sorted.push(left[i++]);
      } else {
        sorted.push(right[j++]);
      }

      arrayAccess += 1; // Accessing an element from left or right
      setArrayAccesses(arrayAccess);

      // Increment the update counter and throttle updates
      updateCounter++;
      if (updateCounter % updateSkip === 0) {
        // Update only the part that changes
        updateArray(prevArray => [
          ...prevArray.slice(0, leftStartIndex),
          ...sorted,
          ...left.slice(i),
          ...right.slice(j),
          ...prevArray.slice(leftStartIndex + left.length + right.length)
        ]);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
    }

    // Add remaining elements
    sorted = [...sorted, ...left.slice(i), ...right.slice(j)];

    // Track accesses for remaining elements
    arrayAccess += left.slice(i).length + right.slice(j).length;
    setArrayAccesses(arrayAccess);

    // Final update with the merged sorted array
    updateArray(prevArray => [
      ...prevArray.slice(0, leftStartIndex),
      ...sorted,
      ...prevArray.slice(leftStartIndex + left.length + right.length)
    ]);

    return sorted;
  };

  const mergeSortHelper = async (arr, startIndex) => {
    if (arr.length < 2) return arr;

    const mid = Math.floor(arr.length / 2);
    const left = await mergeSortHelper(arr.slice(0, mid), startIndex);
    const right = await mergeSortHelper(arr.slice(mid), startIndex + mid);
    return merge(left, right, startIndex);
  };

  // Start the sorting process
  await mergeSortHelper(array, 0);
  highlightIndices([-1, -1]); // Clear highlights

  setSortTime(performance.now() - start);
};

export default mergeSort;

File: backend/api/utils/merge_sort.py
import asyncio

async def merge(arr, left, middle, right, send_fn, delay, throttle, stats):
    comparisons, arrayAccesses, swaps = stats
    n1 = middle - left + 1
    n2 = right - middle

    L = arr[left:middle + 1]
    R = arr[middle + 1:right + 1]
    arrayAccesses += n1 + n2

    i = j = 0  
    k = left  

    while i < n1 and j < n2:
        comparisons += 1
        arrayAccesses += 2
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
            swaps += 1
        k += 1
        arrayAccesses += 1

        if k % throttle == 0:
            await send_fn(arr, comparisons, arrayAccesses, swaps, [k])
            await asyncio.sleep(delay)

    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1
        arrayAccesses += 1

        if k % throttle == 0:
            await send_fn(arr, comparisons, arrayAccesses, swaps, [k])
            await asyncio.sleep(delay)

    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1
        arrayAccesses += 1

        if k % throttle == 0:
            await send_fn(arr, comparisons, arrayAccesses, swaps, [k])
            await asyncio.sleep(delay)

    stats[0], stats[1], stats[2] = comparisons, arrayAccesses, swaps


async def merge_sort_recursive(arr, left, right, send_fn, delay, throttle, stats):
    if left < right:
        middle = (left + right) // 2

        await merge_sort_recursive(arr, left, middle, send_fn, delay, throttle, stats)
        await merge_sort_recursive(arr, middle + 1, right, send_fn, delay, throttle, stats)

        await merge(arr, left, middle, right, send_fn, delay, throttle, stats)


async def merge_sort(arr, send_fn, delay, throttle):
    comparisons = 0
    arrayAccesses = 0
    swaps = 0
    stats = [comparisons, arrayAccesses, swaps]

    await merge_sort_recursive(arr, 0, len(arr) - 1, send_fn, delay, throttle, stats)

    comparisons, arrayAccesses, swaps = stats
    return arr, comparisons, arrayAccesses, swaps
